### PR TITLE
Add TF DeiT implementation

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Adds the TF implementations of DeiT

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by NielsRogge** [comment]
Please also incorporate the updates of #17731

**Comment by amyeroberts** [comment]
> Please also incorporate the updates of https://github.com/huggingface/transformers/pull/17731

@NielsRogge Will do! I originally added, but it resulted in lots of changes because of all the `Copied From` statements, so will wait until your PR is merged and those updates finalised.

File: tests/models/deit/test_modeling_tf_deit.py
Comment (amyeroberts) [inline tests/models/deit/test_modeling_tf_deit.py:272]: To be removed after weights pushed to hub

File: src/transformers/models/deit/modeling_tf_deit.py
Comment (amyeroberts) [inline src/transformers/models/deit/modeling_tf_deit.py:897]: To be removed after weights pushed to hub

File: tests/models/deit/test_modeling_tf_deit.py
Comment (amyeroberts) [inline tests/models/deit/test_modeling_tf_deit.py:248]: To be removed after weights pushed to hub

**Review by Rocketknight1 — APPROVED** [review]
I couldn't see any issues with the ported TF code anywhere! (Except my usual nitpicks that people are using `tf.multiply` and `tf.matmul` instead of just `*` and `@`, but that's purely cosmetic and copied from another file anyway)

Also, tests seem comprehensive, so this PR looks good to me!

**Review by patrickvonplaten — COMMENTED** [review]
Deferring to @NielsRogge and @gante here as it's Vision and TF. Happy to do a final review though if you want :-)

File: src/transformers/models/deit/modeling_tf_deit.py
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_tf_deit.py:2]:
```suggestion
# Copyright 2022 Facebook AI Research (FAIR), Ross Wightman, The HuggingFace Inc. team. All rights reserved.
```

File: src/transformers/models/deit/modeling_tf_deit.py
Comment (NielsRogge) [inline src/transformers/models/deit/modeling_tf_deit.py:938]: Usually I add model output classes at the top of the modeling file

**Review by NielsRogge — APPROVED** [review]
LGTM! Very clean!

File: tests/models/deit/test_modeling_tf_deit.py
Comment (sgugger) [inline tests/models/deit/test_modeling_tf_deit.py:163]:
```suggestion
        config, pixel_values, labels = config_and_inputs
```

**Review by sgugger — APPROVED** [review]
Very clean, thanks a lot for porting this model to TensorFlow!

File: src/transformers/models/deit/modeling_tf_deit.py
Comment (amyeroberts) [reply src/transformers/models/deit/modeling_tf_deit.py]: Agreed - it's much nicer. I've moved it to the top :)

**Pull request merged**
